proposals/0035-linalg-matrix.md

```diff
@@ -47,70 +47,816 @@ concept ArithmeticScalar = std::is_arithmetic<T>::value;
+
+  template <typename T>
+    requires ArithmeticScalar<T>
+  void Store(groupshared T Arr[], uint StartIdx, uint Stride,
+             bool ColMajor);
+
+  // Row accesses
+  vector<ElementType, M> GetRow(uint Index);
+  void SetRow(vector<ElementType, M> V, uint Index);
+
+  // Element access
+  std::enable_if_t<__detail::ComponentTypeTraits<ComponentTy>::IsNativeScalar,
+                   ElementType>
+  Get(uint2 Index);
+  std::enable_if_t<__detail::ComponentTypeTraits<ComponentTy>::IsNativeScalar,
+                   void>
+  Set(ElementType V, uint2 Index);
 
-  void Store(groupshared ElTy Arr[], uint StartIdx, uint Stride, bool ColMajor);
+  template <MatrixComponentType LHSTy, MatrixComponentType RHSTy, unit K>
+  std::enable_if_t<Use == MatrixUse::Accumulator && Scope == MatrixScope::Wave,
+                   void>
+  MultiplyAccumulate(const Matrix<LHSTy, M, K, MatrixUse::A, Scope>,
+                     const Matrix<RHSTy, K, N, MatrixUse::B, Scope>);
 
-  void MultiplyAccumulate(const ref Matrix<T, N, M>);
-  void SumAccumulate(const ref Matrix<T, N, M>);
+  template <MatrixComponentType LHSTy, MatrixComponentType RHSTy, unit K>
+  std::enable_if_t<Use == MatrixUse::Accumulator && Scope == MatrixScope::Wave,
+                   void>
+  SumAccumulate(const Matrix<LHSTy, M, K, MatrixUse::A, Scope>,
+                const Matrix<RHSTy, K, N, MatrixUse::B, Scope>);
+
+  // Cooperative Vector outer product accumulate.
+  template <typename T>
+  std::enable_if_t<Use == MatrixUse::Accumulator, void>
+  OuterProductAccumulate(const vector<T, M> &, const vector<T, N> &);
 };
 
 template <typename T, uint M, uint N, uint K>
-Matrix<T, M, N> Multiply(const ref Matrix<T, M, K>, const ref Matrix<T, K, N>);
+Matrix<T, M, N, MatrixUse::Accumulator, MatrixScope::Wave>
+
+Multiply(const Matrix<T, M, K, MatrixUse::A, MatrixScope::Wave>,
+         const Matrix<T, K, N, MatrixUse::B, MatrixScope::Wave>);
+
+// Cooperative Vector Replacement API
+// Cooperative Vector operates on per-thread vectors multiplying against B
+// matrices.
+
+template <typename OutputElTy, typename InputElTy, uint M, uint K,
+          MatrixComponentType MatrixDT, MatrixScope Scope, bool MatrixTranspose>
+vector<OutputElTy, K>
+Multiply(vector<InputElTy, M> InputVector,
+         Matrix<MatrixDT, M, K, MatrixUse::B, Scope> Matrix);
+
+template <typename OutputElTy, typename InputElTy, typename BiasElTy, uint M,
+          uint K, MatrixComponentType MatrixDT, MatrixScope Scope,
+          bool MatrixTranspose>
+vector<OutputElTy, K>
+MultiplyAdd(vector<InputElTy, M> InputVector,
+            Matrix<MatrixDT, M, K, MatrixUse::B, Scope> Matrix,
+            vector<BiasElTy, K> BiasVector);
 
 } // namespace linalg
 } // namespace hlsl
 ```
 
 ## Detailed design
 
+### HLSL API Concepts
+
+The new HLSL API introduces a new `linalg::Matrix` type which represents an
+opaque matrix object, and contains an intangible handle that refers to the

```

Comment:
What is the sense of intangible here?

The HLSL draft spec says

> standard intangible
types. Intangible types are types that have no defined object representation or value representation, as such the size is
unknown at compile time.

What I would need to know is: what declarations can use a matrix type?
* function param
* function return type
* declaration of a variable/const inside a function
* groupshared variable
* structured buffer, RWstructured buffer
* And can the above use arrays of these matrices, or use the matrix values inside a struct.

Issue [DXC/5935](https://github.com/microsoft/DirectXShaderCompiler/issues/5935) criticizedt he original  wave matrix proposal for not allowing arrays-of-matrices, as used in canonical code samples.